### PR TITLE
Reduce geometric memory growth factor

### PR DESCRIPTION
**Pull request by juj**
Reduce geometric memory growth factor from aggressive 2x to a much more modest +20% factor.

Previous 2x size increase on geometric growth is not really warranted, and the size bumps it causes, mainly 128MB->256MB and 256MB->512MB are quite excessive. This PR changes
 a) the geometric growth rate to be only 20%,
 b) simplifies the program logic around resize,
 c) do not overshoot geometric growth past the requested size, if the requested size already covers the geometric factor,
 d) attempt a more modest growth if a too eager allocation fails.

What is important when considering the growth rate:
1) The number of heap growth operations the application needs to do in the worst case to ramp up to the actual size of memory it is using. Since heap growth can be slow, we do not want to do too many heap grow operations,
2) The growth step in MBytes (marked in parentheses below). The way to think about this number is that it tells how many bytes the application is (in worst case) wasting away as unused memory allocation. We do not want to grow too much, as that causes a lot of waste, and
3) the danger of being too eager to reserve a big memory bump. An application might be able to grow up from 100MB to 150MB it is requesting, but too eager strategy to attempt to shoot directly to a 256MB reservation might fail. The new scheme avoids this by retrying allocation with (d).

This kind of geometric allocation strategy then needs to find a balance between 1) and 2), which are conflicting goals.

After this PR, the worst case geometric memory size increase bumps are
```
initial heap size: x bytes (< 16 MB)
16.00 MB
19.25 MB (+3.25 MB)
23.125 MB (+3.875 MB)
27.75 MB (+4.625 MB)
33.3125 MB (+5.5625 MB)
40 MB (+6.6875 MB)
48 MB (+8 MB)
57.625 MB (+9.625 MB)
69.1875 MB (+11.5625 MB)
83.0625 MB (+13.875 MB)
99.6875 MB (+16.625 MB)
119.625 MB (+19.9375 MB)
143.5625 MB (+23.9375 MB)
172.3125 MB (+28.75 MB)
206.8125 MB (+34.5 MB)
248.1875 MB (+41.375 MB)
297.875 MB (+49.6875 MB)
357.5 MB (+59.625 MB)
429 MB (+71.5 MB)
514.8125 MB (+85.8125 MB)
617.8125 MB (+103 MB)
741.375 MB (+123.5625 MB)
889.6875 MB (+148.3125 MB)
1067.625 MB (+177.9375 MB)
1281.1875 MB (+213.5625 MB)
1537.4375 MB (+256.25 MB)
1844.9375 MB (+307.5 MB)
2047.9375 MB (+203 MB)
```
for a total of 28 increases.

Before this PR, worst case size increases were
```
initial heap size: x bytes (< 16 MB)
32 MB (+16 MB)
64 MB (+32 MB)
128 MB (+64 MB)
256 MB (+128 MB)
512 MB (+256 MB)
896 MB (+384 MB)
1184 MB (+288 MB)
1400 MB (+216 MB)
1562 MB (+162 MB)
1683.5 MB (+121.5 MB)
1774.625 MB (+91.125 MB)
1843 MB (+68.375 MB)
1894.25 MB (+51.25 MB)
1932.6875 MB (+38.4375 MB)
1961.5625 MB (+28.875 MB)
1983.1875 MB (+21.625 MB)
1999.4375 MB (+16.25 MB)
2011.625 MB (+12.1875 MB)
2020.75 MB (+9.125 MB)
2027.5625 MB (+6.8125 MB)
2032.6875 MB (+5.125 MB)
2036.5625 MB (+3.875 MB)
2039.4375 MB (+2.875 MB)
2041.625 MB (+2.1875 MB)
2043.25 MB (+1.625 MB)
2044.4375 MB (+1.1875 MB)
2045.375 MB (+0.9375 MB)
2046.0625 MB (+0.6875 MB)
2046.5625 MB (+0.5 MB)
2046.9375 MB (+0.375 MB)
2047.25 MB (+0.3125 MB)
2047.4375 MB (+0.1875 MB)
2047.625 MB (+0.1875 MB)
2047.75 MB (+0.125 MB)
2047.8125 MB (+0.0625 MB)
2047.875 MB (+0.0625 MB)
2047.9375 MB (+0.0625 MB)
```
By worst case I mean scenarios where heap size would be exhausted and the app would call to request just one more byte beyond the current limit, i.e. the slowest possible growth rate (or max number of grow operations).

Note how the old size limit was very aggressive in initially ramping up the heap sizes by hundreds of megabytes, and after 512MB the growth rate does not much slow down, but the increases go very quickly to around ~1800MB mark. An app is likely to waste memory at the scale of 100MB-300MB with such eager increases, at range of 256MB-1GB of used memory.

The new scheme is much more conservative in saving memory, and wastage is at around ~10MBs range on smaller 256MB sizes, and only grows to 100MB range after 617MB used.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -32,10 +32,10 @@
...
       "requires": {
-        "commander": "2.1.0",
-        "nan": "1.0.0",
-        "options": "0.0.6",
-        "tinycolor": "0.0.1"
+        "commander": "~2.1.0",
+        "nan": "~1.0.0",
+        "options": ">=0.0.5",
+        "tinycolor": "0.x"
       }
```
Comment (kripken) [inline package-lock.json]: why did this file change?

File: src/library.js [hunk 2]
```diff
@@ -560,88 +560,74 @@ LibraryManager.library = {
...
-    if (!replacement) {
-#if ASSERTIONS
-      err('Failed to grow the heap from ' + oldSize + ' bytes to ' + newSize + ' bytes, not enough memory!');
+    var minHeapSize = 16777216;
+
+    // Loop through potential heap size increases. If we attempt a too eager reservation that fails, cut down on the
+    // attempted size and reserve a smaller bump instead. (max 3 times, chosen somewhat arbitrarily)
+    for(var cutDown = 1; cutDown <= 4; cutDown *= 2) {
+#if MEMORY_GROWTH_STEP == -1
+      var heapIncrement = oldSize * (1 + 0.20/cutDown); // ensure geometric growth
```
Comment (kripken) [inline src/library.js]: this doesn't look like an increment, so the name is confusing I think - it's the new size, as it includes both the `1` factor plus the new incremental addition?

File: package-lock.json (hunk 1 above)
Comment (juj) [reply package-lock.json]: Oops, by accident. (I'll commit a fixed package-lock.json once html-minifier and closure npm update PRs land)

File: src/library.js (hunk 2 above)
Comment (juj) [reply src/library.js]: renamed to `overGrownHeapSize`

**Comment by juj** [comment]
> In general this looks good, but I wonder how we can reason about the 20% figure? My intuition is that it's too low,

At first I was going with +15% thinking that is more than enough, though then went for a bit more generous increment.

I think we need to find a balance at the two drawbacks: too large increment wastes hundreds of MBs of memory in form of over-reserving, and too small increment causes "GC-like" or "JIT-like" hiccups when growing. To know the right balance, we should quantify these drawbacks.

The amount of wastage is quantified in above comment. With 2x increment, max overreservation is 384MB when going from 512MB to 896 MB. With +20% increment, max overreservation is 1537.4375 MB (+256.25 MB)
1844.9375 MB (+307.5 MB)

How slow is memory growth then? Testing out with the following program:

`a.c`

```c
#include <stdio.h>
#include <stdlib.h>

int ptr = 0;
int main()
{
	for(int i = 0; i < 1024*1024*1024; ++i) {
		void *p = malloc(1024);
		ptr += (int)p;
		if (!p)
			break;
	}
	printf("%d\n", ptr);
}
```

```
emcc a.c -o a.html -s ALLOW_MEMORY_GROWTH=1 -O3 -s ASSERTIONS=1
```

Testing out, Chrome 79:
```
Heap resize call from 16777216 to 20185088 took 0.2550000208429992 msecs. Success: true
Heap resize call from 20185088 to 24248320 took 0.18000000272877514 msecs. Success: true
Heap resize call from 24248320 to 29097984 took 0.1449999981559813 msecs. Success: true
Heap resize call from 29097984 to 34930688 took 0.20000000949949026 msecs. Success: true
Heap resize call from 34930688 to 41943040 took 0.9300000092480332 msecs. Success: true
Heap resize call from 41943040 to 50331648 took 1.6250000044237822 msecs. Success: true
Heap resize call from 50331648 to 60424192 took 5.545000021811575 msecs. Success: true
Heap resize call from 60424192 to 72548352 took 0.8249999955296516 msecs. Success: true
Heap resize call from 72548352 to 87097344 took 0.8399999933317304 msecs. Success: true
Heap resize call from 87097344 to 104529920 took 3.6100000143051147 msecs. Success: true
Heap resize call from 104529920 to 125435904 took 1.7349999980069697 msecs. Success: true
Heap resize call from 125435904 to 150536192 took 1.8400000117253512 msecs. Success: true
Heap resize call from 150536192 to 180682752 took 3.939999995054677 msecs. Success: true
Heap resize call from 180682752 to 216858624 took 3.005000005941838 msecs. Success: true
Heap resize call from 216858624 to 260243456 took 8.320000022649765 msecs. Success: true
Heap resize call from 260243456 to 312344576 took 4.2449999891687185 msecs. Success: true
Heap resize call from 312344576 to 374865920 took 5.550000001676381 msecs. Success: true
Heap resize call from 374865920 to 449839104 took 10.609999997541308 msecs. Success: true
Heap resize call from 449839104 to 539820032 took 8.269999991171062 msecs. Success: true
Heap resize call from 539820032 to 647823360 took 8.165000006556511 msecs. Success: true
Heap resize call from 647823360 to 777388032 took 14.66499999514781 msecs. Success: true
Heap resize call from 777388032 to 932904960 took 13.135000015608966 msecs. Success: true
Heap resize call from 932904960 to 1119485952 took 13.829999981680885 msecs. Success: true
Heap resize call from 1119485952 to 1343422464 took 21.385000000009313 msecs. Success: true
Heap resize call from 1343422464 to 1612120064 took 21.03999999235384 msecs. Success: true
Heap resize call from 1612120064 to 1934557184 took 22.884999983944 msecs. Success: true
Heap resize call from 1934557184 to 2147418112 took 30.40999997756444 msecs. Success: true
```
Firefox 71:
```
Heap resize call from 16777216 to 20185088 took 0.07800000000000296 msecs. Success: true
Heap resize call from 20185088 to 24248320 took 0.054400000000015325 msecs. Success: true
Heap resize call from 24248320 to 29097984 took 0.04560000000000741 msecs. Success: true
Heap resize call from 29097984 to 34930688 took 0.052199999999999136 msecs. Success: true
Heap resize call from 34930688 to 41943040 took 0.050199999999989586 msecs. Success: true
Heap resize call from 41943040 to 50331648 took 0.04959999999999809 msecs. Success: true
Heap resize call from 50331648 to 60424192 took 2.2335000000000065 msecs. Success: true
Heap resize call from 60424192 to 72548352 took 0.05870000000000175 msecs. Success: true
Heap resize call from 72548352 to 87097344 took 0.06560000000000343 msecs. Success: true
Heap resize call from 87097344 to 104529920 took 1.5259999999999962 msecs. Success: true
Heap resize call from 104529920 to 125435904 took 0.2348999999999961 msecs. Success: true
Heap resize call from 125435904 to 150536192 took 0.08780000000000143 msecs. Success: true
Heap resize call from 150536192 to 180682752 took 1.3166000000000224 msecs. Success: true
Heap resize call from 180682752 to 216858624 took 0.11820000000000164 msecs. Success: true
Heap resize call from 216858624 to 260243456 took 0.1150000000000091 msecs. Success: true
Heap resize call from 260243456 to 312344576 took 1.3583999999999605 msecs. Success: true
Heap resize call from 312344576 to 374865920 took 0.16970000000000596 msecs. Success: true
Heap resize call from 374865920 to 449839104 took 0.18019999999998504 msecs. Success: true
Heap resize call from 449839104 to 539820032 took 1.590299999999985 msecs. Success: true
Heap resize call from 539820032 to 647823360 took 0.23099999999996612 msecs. Success: true
Heap resize call from 647823360 to 777388032 took 0.27880000000004657 msecs. Success: true
Heap resize call from 777388032 to 932904960 took 1.826599999999985 msecs. Success: true
Heap resize call from 932904960 to 1119485952 took 0.3905000000000314 msecs. Success: true
Heap resize call from 1119485952 to 1343422464 took 0.47080000000005384 msecs. Success: true
Heap resize call from 1343422464 to 1612120064 took 1.8410999999998694 msecs. Success: true
Heap resize call from 1612120064 to 1934557184 took 0.7561000000000604 msecs. Success: true
Heap resize call from 1934557184 to 2147418112 took 0.5178000000000793 msecs. Success: true
```
Node.js v12.9.1:
```
Heap resize call from 16777216 to 20185088 took 0.25369900465011597 msecs. Success: true
Heap resize call from 20185088 to 24248320 took 0.15450000762939453 msecs. Success: true
Heap resize call from 24248320 to 29097984 took 0.1184999942779541 msecs. Success: true
Heap resize call from 29097984 to 34930688 took 0.24210000038146973 msecs. Success: true
Heap resize call from 34930688 to 41943040 took 0.6951000094413757 msecs. Success: true
Heap resize call from 41943040 to 50331648 took 8.032298982143402 msecs. Success: true
Heap resize call from 50331648 to 60424192 took 0.586698979139328 msecs. Success: true
Heap resize call from 60424192 to 72548352 took 0.8524999916553497 msecs. Success: true
Heap resize call from 72548352 to 87097344 took 0.938198983669281 msecs. Success: true
Heap resize call from 87097344 to 104529920 took 1.7804009914398193 msecs. Success: true
Heap resize call from 104529920 to 125435904 took 1.5771999955177307 msecs. Success: true
Heap resize call from 125435904 to 150536192 took 12.081999987363815 msecs. Success: true
Heap resize call from 150536192 to 180682752 took 2.0890989899635315 msecs. Success: true
Heap resize call from 180682752 to 216858624 took 2.728899985551834 msecs. Success: true
Heap resize call from 216858624 to 260243456 took 3.076698988676071 msecs. Success: true
Heap resize call from 260243456 to 312344576 took 11.705300003290176 msecs. Success: true
Heap resize call from 312344576 to 374865920 took 4.27020001411438 msecs. Success: true
Heap resize call from 374865920 to 449839104 took 5.861901015043259 msecs. Success: true
Heap resize call from 449839104 to 539820032 took 6.161300003528595 msecs. Success: true
Heap resize call from 539820032 to 647823360 took 15.79229998588562 msecs. Success: true
Heap resize call from 647823360 to 777388032 took 8.31360000371933 msecs. Success: true
Heap resize call from 777388032 to 932904960 took 8.314400017261505 msecs. Success: true
Heap resize call from 932904960 to 1119485952 took 17.185698986053467 msecs. Success: true
Heap resize call from 1119485952 to 1343422464 took 12.333099007606506 msecs. Success: true
Heap resize call from 1343422464 to 1612120064 took 14.256999015808105 msecs. Success: true
Heap resize call from 1612120064 to 1934557184 took 17.912499994039536 msecs. Success: true
Heap resize call from 1934557184 to 2147418112 took 20.504601001739502 msecs. Success: true
```
So it looks like we are talking about one or two missed frames in a game application. In Firefox heap resizes are practically free. Even for Unity's needs, a few missed frames is gladly welcome if that means a memory save of 100MB overall.

Repeating the same test in a pthreads build with

```
emcc a.c -o a.html -s ALLOW_MEMORY_GROWTH=1 -O3 -s ASSERTIONS=1 -s USE_PTHREADS=1 -s WASM_MEM_MAX=2147418112
```
yielded very similar numbers in Chrome.

Would we be willing to have each Emscripten compiled page waste up to 384MB of browser's memory, or should we cap the waste to e.g. 100MB at most, and introduce 3x extra heap resizes (á 20msecs each?). Even for a high speed game application I think the answer is a clear yes.

Actually I think the answer is so clear yes, that I added a max 96MB cap to the grow operations to strangle the heap wastage running loose at heap sizes beyond 512MB.

Very much recommend that by default the compiler should should **side towards **extremely aggressive memory savings**, and err towards adding possibly 30x 20 msec hiccups to each built page, rather than have each built page consume between 100MB-400MB of extra unused memory.

People have different needs, and it is possible that e.g. a +20% geometric rate with a 96MB cap will not suit everyone's needs. I think I'd like to have someone actually run into issues with the 20 msec hiccups, and complain (which will prompt Emscripten perhaps to add a linker option to customize the geometric growth rate). That would be much better than having all WebAssembly pages out in the world by default waste ~100MB-400MB range of memory? (and then have non-developers try to complain to developers why Wasm pages consume so much memory)

**Comment by juj** [comment]
> d) attempt a more modest growth if a too eager allocation fails.

A special note here is that the eager allocation can also crash the whole tab, can happen at least on Safari, perhaps also in Chrome (although for Chrome it is only hearsay, no repro known)

**Comment by juj** [comment]
Actually, let me go ahead and add the option to adjust the growth rates already now, since it is foreseeable that some people may want to configure it (and adjusting it is just a matter of changing a single value, so very self-contained).

How does this look now?

File: tests/test_other.py [hunk 3]
```diff
@@ -6157,7 +6157,10 @@ def test_massive_alloc(self):
     run_process([PYTHON, EMCC, 'main.cpp', '-s', 'ALLOW_MEMORY_GROWTH=1', '-s', 'WASM=0'])
     # just care about message regarding allocating over 1GB of memory
     output = run_js('a.out.js', stderr=PIPE, full_output=True)
-    self.assertContained('''Warning: Enlarging memory arrays, this is not fast! 16777216,1543503872\n''', output)
+    if self.is_wasm_backend():
+      self.assertContained('''Warning: Enlarging memory arrays, this is not fast! 16777216,1473314816\n''', output)
+    else:
+      self.assertContained('''Warning: Enlarging memory arrays, this is not fast! 16777216,1476395008\n''', output)
```
Comment (kripken) [inline tests/test_other.py]: why is this now different between the backends?

**Comment by VirtualTim** [comment]
Might be good to wait for #10052 to land first?

File: tests/test_other.py (hunk 3 above)
Comment (juj) [reply tests/test_other.py]: I think wasm2js has page size of 64KB, asm.js has page size of 16MB.

**Pull request merged**